### PR TITLE
Parameter manager

### PR DESCRIPTION
**Pull request by dirkpr**
✨ Refactor Parameter Management System and Enhance Flexibility
Extends: #93
Overview
This PR refactors the parameter management system to provide more intuitive and flexible handling of learnable parameters across optimization stages, while maintaining backward compatibility.

Key Changes
🔧 Enhanced Stage-Level Parameter Control

Introduce vary_stages field to specify exactly which stages a parameter should change at
Enable fine-grained control over parameter variation across multi-stage optimization problems

🏗️ Improved API Design

Replace ambiguous field names (differentiable, fix) with clear, semantic interface terms:

learnable - parameter can be optimized
non_learnable - parameter is constant but accessible
fix - parameter is locked to a specific value

📊 Matrix Parameter Support

Add support for matrix-valued parameters with per-element bounds
Implement intelligent bound handling with automatic -/+ infinity defaults for undefined bounds

✅ Comprehensive Testing

Expand test suite to cover new functionality
Add validation for edge cases and boundary conditions

Benefits

More intuitive parameter configuration with self-documenting field names
Greater flexibility in defining parameter behavior across optimization stages
Better support for complex parameter types (matrices with element-wise constraints)
Improved developer experience with clearer API semantics

Notes
This change includes API improvements that require updating existing parameter configurations to use the new field names.

### CODE REVIEW

File: leap_c/parameters.py [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+    upper_bound: np.ndarray | None = None
+    interface: Literal["fix", "learnable", "non-learnable"] = "fix"
+
+
+class ParameterManager:
+    """Manager for parameters."""
+
+    parameters: dict[str, Parameter] = {}
+    # TODO: Check if type hint is correct
+    learnable_parameters: dict[str, dict[str, int | np.ndarray]] = {}
```
Comment (JasperHoffmann) [inline leap_c/parameters.py:36]: Is the type hint not enough? Maybe remove `= {}`

File: leap_c/parameters.py [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+        self,
+        batch_size: int | None = None,
+        **overwrite: np.ndarray,
+    ) -> np.ndarray:
+        """
+        Combine all non-learnable parameters into a single numpy array.
+
+        Args:
+            batch_size: The batch size for the parameters.
+            Not needed if overwrite is provided.
```
Comment (JasperHoffmann) [inline leap_c/parameters.py:159]: Indent correctly?

File: leap_c/parameters.py [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+    def learnable_params_default(self) -> np.ndarray:
+        """
+        Return the default values for all learnable parameters.
+
+        Returns:
+            np.ndarray: Flattened array of default values for learnable parameters.
+        """
+        return self.learnable_array.copy()
+
+    def learnable_params_lower_bound(self) -> np.ndarray:
```
[the remark this answers is not in the document]
Comment (dirkpr) [reply leap_c/parameters.py:214]: I am not sure. We are already seeing cases where people do not use gym.Env. And I would use the bounds to parameterize the torch.distribution to sample from.

File: tests/leap_c/ocp/acados/test_acados_parameters.py [hunk 4]
```diff
@@ -0,0 +1,816 @@
...
+    assert "price_3_6" in price_keys
+    assert "price_7_10" in price_keys
+
+    # demand changes at [2, 5, 8], so we expect: demand_0_1, demand_2_4, demand_5_7, demand_8_10
+    demand_keys = [k for k in learnable_keys if k.startswith("demand_")]
+    assert len(demand_keys) == 4
+    assert "demand_0_1" in demand_keys
+    assert "demand_2_4" in demand_keys
+    assert "demand_5_7" in demand_keys
+    assert "demand_8_10" in demand_keys
```
[the remark this answers is not in the document]
Comment (dirkpr) [reply tests/leap_c/ocp/acados/test_acados_parameters.py:112]: The motivation for this test is to check that the param manager is creating the symbols according to the specification in vary_stages

File: leap_c/parameters.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+        # Build map for non-learnable parameters and create flattened array
+        self.non_learnable_parameters = {}
+        non_learnable_arrays = []
+        current_index = 0
+
+        for name, parameter in self.parameters.items():
+            if parameter.interface == "non-learnable":
+                param_size = parameter.default.size
+                self.non_learnable_parameters[name] = {
+                    "start_idx": current_index,
```
[the remark this answers is not in the document]
Comment (dirkpr) [reply leap_c/parameters.py:74]: I do not see how this is specific to `acados`? In the current implementation, the indices and shape are stored to convert between a flat array and specific elements.

File: leap_c/parameters.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+                    "end_idx": current_index + param_size,
+                    "shape": parameter.default.shape,
+                }
+                non_learnable_arrays.append(parameter.default.flatten())
+                current_index += param_size
+
+        self.non_learnable_array = (
```
[the remark this answers is not in the document]
Comment (dirkpr) [reply leap_c/parameters.py]: Removed it

**Review by Confectio — CHANGES_REQUESTED** [review]
Nice :) . I still have some questions and some comments

File: leap_c/examples/cartpole/acados_ocp.py [hunk 7]
```diff
@@ -14,6 +14,7 @@
 
 def create_cartpole_params(
     param_interface: CartPoleAcadosParamInterface,
+    N_horizon: int = 50,
```
[the remark this answers is not in the document]
Comment (dirkpr) [reply leap_c/examples/cartpole/acados_ocp.py:17]: This is the default from the example: https://github.com/leap-c/leap-c/blob/8df21285d269a35c297a337ccae2d79d28435d1a/leap_c/examples/cartpole/acados_ocp.py#L87

File: leap_c/ocp/acados/parameters.py [hunk 8]
```diff
@@ -21,210 +23,175 @@ class Parameter(NamedTuple):
...
-        """Get all differentiable stage-wise parameters."""
-        return {
-            key: value.value
-            for key, value in self.parameters.items()
-            if value.differentiable and value.stagewise
-        }
+        # Now build the lower and upper bound
+        self.learnable_parameters_default = self.learnable_parameters(0)
+        self.learnable_parameters_lb = self.learnable_parameters(0)
+        self.learnable_parameters_ub = self.learnable_parameters(0)
```
[the remark this answers is not in the document]
Comment (dirkpr) [reply leap_c/ocp/acados/parameters.py:144]: It creates a `casadi.tools.structure3.DMStruct` from the `casadi.tools.structure3.ssymStruct` (here `self.learnable_parameters`) with the same structure but all entries are zero and type `DM`. In the code we then insert the correct values.

File: leap_c/ocp/acados/parameters.py [hunk 9]
```diff
@@ -270,78 +238,92 @@ def combine_parameter_values(
...
 
-        if field in self._get_differentiable_stagewise_parameters():
-            return sum(
-                [
-                    self.p["indicator"][stage_] * self.p_global[field, stage_]
-                    for stage_ in range(self.N_horizon + 1)
-                ]
+        if (
+            self.parameters[field].interface == "learnable"
+            and self.parameters[field].vary_stages
```
[the remark this answers is not in the document]
Comment (dirkpr) [reply leap_c/ocp/acados/parameters.py:281]: Not sure if I understand the nuance between the two cases you describe. But the idea is that by using a separate symbol in `p_global` for each stage and multiplying it with the corresponding indicator variable in `p`, we build stagewise differentiable parameters. The non-learnable parameters, which enter the `AcadosOcp` through `p` have a different interface.

File: leap_c/parameters.py [hunk 10]
```diff
@@ -0,0 +1,114 @@
+from typing import NamedTuple, Literal
+import numpy as np
+import gymnasium as gym
+
+
+class Parameter(NamedTuple):
```
[the remark this answers is not in the document]
Comment (dirkpr) [reply leap_c/parameters.py:6]: Implemented a workaround for inheritance from `NamedTuple` here [445c7a4](https://github.com/leap-c/leap-c/pull/108/commits/445c7a4925c3b6f90f5ad08c34197504e9585715). Not sure when we would need it though.

File: leap_c/ocp/acados/parameters.py [hunk 11]
```diff
@@ -1,9 +1,11 @@
...
+import warnings
 
 import casadi as ca
 import numpy as np
 from acados_template import AcadosOcp
 from casadi.tools import entry, struct, struct_symSX
+import gymnasium as gym
 
 
 class Parameter(NamedTuple):
```
[the remark this answers is not in the document]
Comment (dirkpr) [reply leap_c/ocp/acados/parameters.py]: Also addressed with [445c7a4](https://github.com/leap-c/leap-c/pull/108/commits/445c7a4925c3b6f90f5ad08c34197504e9585715)

File: tests/leap_c/ocp/acados/test_acados_parameters.py [hunk 12]
```diff
@@ -0,0 +1,1004 @@
...
+
+    with pytest.raises(
+        ValueError,
+        match="Parameter 'tensor_upper_bounds_param' upper_bound has 3 dimensions.*CasADi only supports arrays up to 2 dimensions",
+    ):
+        AcadosParameterManager(params_3d_upper_bounds, N_horizon=5)
+
+
+def test_combine_parameter_values():
+    """Test that combine_parameter_values method exists but may not be fully implemented."""
```
[the remark this answers is not in the document]
Comment (dirkpr) [reply tests/leap_c/ocp/acados/test_acados_parameters.py]: That docstring was old. Updated it with https://github.com/leap-c/leap-c/pull/108/commits/25cba8a8bed10c470f8fc6cc4fef45a1edeccb56

File: leap_c/ocp/acados/parameters.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Confectio) [reply leap_c/ocp/acados/parameters.py:340]: When looking at the function signature, a ``field`` is given, and a ``stage`` can be given. In the case ``stage`` is not given, I expect what the function does right now is the desired result for ``field="learnable"``. However, if ``stage`` is also given, I would expect to only return the variables corresponding to the given ``stage`` (maybe multiplied with the respective indicators). Or to put it differently: The function input ``stage`` is not used for ``field=learnable``.

But now that I think about it, I guess nobody has to use this functionality anyways, and can just use the whole variable as returned here, since this variable is made stagewise by the indicators at all times. 
The only improvement I can think of would be a cleaner signature, in case someone has to inspect their casadi expression by string :) .
I think mentioning this somewhere, probably in the docstring would be nice, but I can do this in the docs PR.

File: tests/leap_c/ocp/acados/test_acados_parameters.py [hunk 13]
```diff
@@ -0,0 +1,1004 @@
...
+
+    Args:
+        acados_test_ocp_with_stagewise_varying_params: AcadosOcp instance
+        with stagewise varying parameters.
+        nominal_stagewise_params: Tuple of
+         test parameters to overwrite.
+        rng: Random number generator for reproducible noise.
+
+    Raises:
+        AssertionError: If the mapped and retrieved dense values do not match within
```
[the remark this answers is not in the document]
Comment (Confectio) [reply tests/leap_c/ocp/acados/test_acados_parameters.py:870]: Maybe Im blind, but I dont see them? :D I only see one assertion in line 845?

File: leap_c/examples/hvac/config.py [hunk 14]
```diff
@@ -176,30 +178,24 @@ def make_default_hvac_params(stagewise: bool = False) -> tuple[AcadosParameter,
...
-                value=np.array([0.001]),  # weight on rate of change of heater power
-                lower_bound=np.array([0.0001]),
-                upper_bound=np.array([0.001]),
-                fix=False,
-                differentiable=True,
-                stagewise=stagewise,
+                default=np.array([0.001]),  # weight on rate of change of heater power
+                space=gym.spaces.Box(low=np.array([0.0001]), high=np.array([0.001]), dtype=np.float64),
+                interface="learnable",
+                vary_stages=list(range(48)) if stagewise else [],
```
Comment (Confectio) [inline leap_c/examples/hvac/config.py:184]: The ``48`` should depend on the horizon, no? Also below

**Review by Confectio — APPROVED** [review]
